pkg/syncer/sync/reconciler.go

```diff
@@ -188,29 +192,22 @@ func (r *metaReconciler) reconcileSyncs(ctx context.Context, request reconcile.R
 }
 
 func (r *metaReconciler) finalizeSync(ctx context.Context, sync *v1.Sync, gvks map[schema.GroupVersionKind]bool) status.MultiError {
-	var newFinalizers []string
-	var needsFinalize bool
-	for _, f := range sync.Finalizers {
-		if f == v1.SyncFinalizer {
-			needsFinalize = true
-		} else {
-			newFinalizers = append(newFinalizers, f)
-		}
-	}
-
 	// Check if Syncer finalizer is present before finalize.
-	if !needsFinalize {
+	if !controllerutil.ContainsFinalizer(sync, v1.SyncFinalizer) {
 		klog.V(2).Infof("Sync %s already finalized", sync.Name)
 		return nil
 	}
 
-	sync = sync.DeepCopy()
-	sync.Finalizers = newFinalizers
 	klog.Infof("beginning Sync finalize for %s", sync.Name)
 	if err := r.gcResources(ctx, sync, gvks); err != nil {
 		return err
 	}
-	err := r.client.Upsert(ctx, sync)
+	_, err := r.client.Apply(ctx, sync, func(obj client.Object) (client.Object, error) {
```

Comment:
Should this be `Apply` or `Update`?

Reply:
I changed it to an Apply because we don't actually care about the original state of the Sync. All we want to do is remove the finalizer, if it still exists. Plus, since the Sync was last Get before a bunch of Deletes in gcResources, we want to do an Apply instead of just an Update to avoid conflicting on any sort of drift that's happened since the Get.

Reply:
That makes sense.